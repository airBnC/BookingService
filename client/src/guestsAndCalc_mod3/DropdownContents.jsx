import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const GuestHolder = styled.div`
  display: flex;
  flex-direction: row;
`;

const GuestDetails = styled.div`
`;

const DownButton = styled.button`
`;

const UpButton = styled.button`
`;

export default function DropdownContents( props ) {
  DropdownContents.propTypes = {
    adults: PropTypes.number,
    childrens: PropTypes.number,
    infants: PropTypes.number,
    btnClick: PropTypes.func,
  };
  DropdownContents.defaultProps = {
    adults: 0,
    childrens: 0,
    infants: 0,
    btnClick: () => null,
  };

  const thingsToRender = [ 'adults', 'children', 'infants' ];
  const toRender = thingsToRender.map( thing =>
    ( <GuestHolder>
      <div> {thing} </div>
      <DownButton />
      <div> {props.adults} </div>
      <UpButton />
      </GuestHolder> ) );

  return (
    <div>
      {toRender}
    </div>

  );
}
