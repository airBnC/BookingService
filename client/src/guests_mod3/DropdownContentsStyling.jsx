import styled from 'styled-components';
import { abnbGrey, abnbDarkBlue } from '../ModStylings';

export const Holder = styled.div`
  display: flex;
  flex-direction: row;
  margin-top: 10px;
  justify-content: space-between;
`;

export const Guest = styled.div`
  text-transform: capitalize;
  margin-top: ${props => (props.guest === 'adults' ? '10px' : '0px')};
  font-weight: bold;
`;

export const GuestDetails = styled.div`
  font-size: 12px;
  min-height: 12px;
`;

export const FeatureHolder = styled.div`
  display: flex;
  flex-direction: ${props => ((props.button) ? 'row' : 'column')};
  margin-top: ${props => (props.guest === 'adults' ? '4px' : '0px')}
`;

export const Number = styled.div`
  min-width: 9px;
  font-size: 17px;
  margin-top: 2px;
`;

export const Button = styled.button`
  border-radius: 50%;
  height: 25px;
  width: 25px;
  margin-left: 7px;
  margin-right: 7px;
  margin-top: 1px;
  outline: none;
    &:active{
      background-color: ${abnbDarkBlue};
    }
    &:hover:enabled{
      cursor: pointer;
    }
`;

export const ButtonDown = Button.extend`
  border: ${(props) => {
    if (props.number > 0 && (props.guest === 'infants' || props.guest === 'children')) {
      return `1px solid ${abnbDarkBlue}`;
    } else if (props.guest === 'adults' && props.number > 1) {
      return `1px solid ${abnbDarkBlue}`;
    }
    return `1px solid ${abnbGrey}`;
  }};
`;

export const ButtonUp = Button.extend`
  border: ${(props) => {
    if (props.total < props.max && (props.guest === 'adults' || props.guest === 'children')) {
      return `1px solid ${abnbDarkBlue}`;
    } else if (props.guest === 'infants' && props.number < 5) {
      return `1px solid ${abnbDarkBlue}`;
    }
    return `1px solid ${abnbGrey}`;
  }};
`;
