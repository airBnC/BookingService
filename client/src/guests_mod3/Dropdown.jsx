import React from 'react';
import PropTypes from 'prop-types';
import DropdownContents from './DropdownContents';
import { DropdownHolder, Close, Details } from '../ModStylings';

export default function Dropdown(props) {
  if (props.clicked) {
    return (
      <DropdownHolder guest>
        <DropdownContents
          {...props}
        />
        <Details>
          {props.maxGuests } guests maximum.
          Infants donâ€™t count toward the number of guests.
        </Details>
        <Close onClick={props.close}> Close </Close>
      </DropdownHolder>
    );
  }
  return null;
}

Dropdown.propTypes = {
  clicked: PropTypes.bool,
  btnClick: PropTypes.func,
  close: PropTypes.func,
  adult: PropTypes.number,
  child: PropTypes.number,
  infant: PropTypes.number,
  totalGuests: PropTypes.number,
  maxGuests: PropTypes.number,

};
Dropdown.defaultProps = {
  clicked: false,
  btnClick: () => null,
  close: () => null,
  adult: 0,
  child: 0,
  infant: 0,
  totalGuests: 0,
  maxGuests: 0,
};
