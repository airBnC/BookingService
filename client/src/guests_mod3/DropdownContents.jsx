import React from 'react';
import PropTypes from 'prop-types';
import { Holder, FeatureHolder, Guest, GuestDetails, ButtonDown, ButtonUp, Number } from './DropdownContentsStyling';

const list = ['adults', 'children', 'infants'];
const desc = ['', 'Ages 2-12', 'Under 2'];

export default function DropdownContents(props) {
  const toRender = list.map((propName, index) => {
    const guest = propName.replace(/s|ren/g, '');
    const guestNum = props[guest];
    return (
      <Holder key={propName}>
        <FeatureHolder>
          <Guest guest={propName}> {propName} </Guest>
          <GuestDetails> {desc[index]} </GuestDetails>
        </FeatureHolder>
        <FeatureHolder button guest={propName}>
          <ButtonDown
            disabled={(guest === 'adult' && guestNum === 1) || guestNum === 0}
            onClick={() => props.btnClick(propName, -1)}
            number={guestNum}
            guest={propName}
          >-
          </ButtonDown>
          <Number> {guestNum} </Number>
          <ButtonUp
            disabled={(guest === 'infant' && guestNum === 5) || (props.totalGuests === props.maxGuests && (guest === 'adult' || guest === 'child'))}
            onClick={() => props.btnClick(propName, 1)}
            number={guestNum}
            total={props.totalGuests}
            max={props.maxGuests}
            guest={propName}
          >+
          </ButtonUp>
        </FeatureHolder>
      </Holder>
    );
  });

  return (
    <div>
      {toRender}
    </div>

  );
}

DropdownContents.propTypes = {
  // btnClick: PropTypes.func,
  // adult: PropTypes.number,
  // child: PropTypes.number,
  // infant: PropTypes.number,
  totalGuests: PropTypes.number,
  maxGuests: PropTypes.number,

};
DropdownContents.defaultProps = {
  // btnClick: () => null,
  // adult: 0,
  // child: 0,
  // infant: 0,
  totalGuests: 0,
  maxGuests: 0,
};
