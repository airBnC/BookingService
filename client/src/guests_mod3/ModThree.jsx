import React from 'react';
import PropTypes from 'prop-types';
import Dropdown from './Dropdown';
import { Holder, InputHolder, Title, Logo, Contents } from '../ModStylings';

export default function ModThree(props) {
  let infantStatement;
  if (props.infant > 1) {
    infantStatement = `, ${props.infant} infants`;
  } else if (props.infant === 1) {
    infantStatement = ', 1 infant';
  }

  return (
    <Holder>
      <Title> Guests </Title>
      <InputHolder onClick={props.close}>
        <Contents clicked={props.clicked} mod3>
          {props.totalGuests} {props.totalGuests > 1 ? 'guests' : 'guest'}
          {infantStatement}
        </Contents>
        <Logo
          clicked={props.clicked}
        />
      </InputHolder>
      <Dropdown
        clicked={props.clicked}
        {...props}
      />

    </Holder>
  );
}

ModThree.propTypes = {
  btnClick: PropTypes.func,
  close: PropTypes.func,
  clicked: PropTypes.bool,
  adult: PropTypes.number,
  child: PropTypes.number,
  infant: PropTypes.number,
  totalGuests: PropTypes.number,
  maxGuests: PropTypes.number,
};
ModThree.defaultProps = {
  btnClick: () => null,
  close: () => null,
  clicked: false,
  adult: 0,
  child: 0,
  infant: 0,
  totalGuests: 0,
  maxGuests: 0,
};
