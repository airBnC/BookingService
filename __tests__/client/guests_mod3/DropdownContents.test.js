import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import 'jest-styled-components';
import { configure, shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import sinon from 'sinon';
import DropdownContents from '../../../client/src/guests_mod3/DropdownContents';
import { Holder, FeatureHolder, Guest, GuestDetails, ButtonDown, ButtonUp, Number } from '../../../client/src/guests_mod3/DropdownContentsStyling';

configure({ adapter: new Adapter() });

const props = {
  btnClick: () => null,
  adult: 1,
  child: 0,
  infant: 0,
  totalGuests: 1,
  maxGuests: 6,
};

describe('Holder Component', () => {
  it('is the main flex component', () => {
    const tree = renderer
      .create(<Holder> things inside </Holder>)
      .toJSON();
    expect(tree).toHaveStyleRule('display', 'flex');
    expect(tree).toMatchSnapshot();
  });
});
describe('Dropdown Contents Component', () => {
  it('renders the adults, children, infants, and their descriptions', () => {
    const wrapper = shallow(<DropdownContents {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Dropdown Contents Buttons', () => {
  let subtract;
  let btn;
  beforeEach(() => {
    subtract = jest.fn();
    btn = mount(<ButtonDown onClick={subtract} />);
  });
  it('ButtonDown requires subtract func', () => {
    expect(btn.props().onClick).toBeDefined();
  });
  it('ButtonDown renders a button', () => {
    const button = btn.find('button').first();
    expect(button).toBeDefined();
  });
  it('ButtonDown click calls subtract', () => {
    subtract = sinon.spy();
    const wrapper = mount(<ButtonDown onClick={subtract} />);
    wrapper.find('button').simulate('click');
    expect(subtract.calledOnce).toEqual(true);
  });

  let add;
  beforeEach(() => {
    add = jest.fn();
    btn = mount(<ButtonDown onClick={add} />);
  });
  it('ButtonUp requires subtract func', () => {
    expect(btn.props().onClick).toBeDefined();
  });
  it('ButtonUp renders a button', () => {
    const button = btn.find('button').first();
    expect(button).toBeDefined();
  });
  it('ButtonDown click calls add', () => {
    add = sinon.spy();
    const wrapper = mount(<ButtonDown onClick={add} />);
    wrapper.find('button').simulate('click');
    expect(add.calledOnce).toEqual(true);
  });
});
