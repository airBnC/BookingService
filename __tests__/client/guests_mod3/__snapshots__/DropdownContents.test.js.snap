// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Contents Component renders the adults, children, infants, and their descriptions 1`] = `
Array [
  <DropdownContentsStyling__Holder
    key="adults"
  />,
  <DropdownContentsStyling__Holder
    key="children"
  />,
  <DropdownContentsStyling__Holder
    key="infants"
  />,
]
`;

exports[`Holder Component is the main flex component 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  margin-top: 10px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

<div
  className="c0"
>
   things inside 
</div>
`;
